Fixing link component

josephsintum opened the pull request:
#86 

Making a custom link component that determines internally if it is a internal or external link and automatically uses the correct tag and correct styling and avoid future errors

`touch /src/Link.tsx`

```tsx
import React from "react";
import { HashLink, HashLinkProps } from "react-router-hash-link";
import { Link as MaterialLink } from "@material-ui/core";

const InternalLink = React.forwardRef((props: HashLinkProps, ref) => (
  <HashLink {...props} />
));

export const Link = (props) => {
  if (props.to.indexOf("://")=== -1) {
    return <MaterialLink component={InternalLink} {...props}/>
  } else {
    return <MaterialLink {...props} component="a" href={props.to} />
  }
}
```


then usage is really simple

```tsx
import React from "react";
import { Link } from "./Links";

export const RandomComponent = (props) => {
  <>
      <Link to="/benchmarks">Styled Link</Link>
      <Link to="https://github.com/denoland/deno">master branch</Link>
  </>
}```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/deno_website2?pullRequest=91) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/denoland/deno_website2?pullRequest=91) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/deno_website2?pullRequest=91) it.</sub>

lucacasonato commented:
Looks good!

What do you think about also adding a `<Button>` component that uses MaterialButton? That way we get the same improvements for buttons too, not just links.

Oh and it would be cool if you could also add some tests :-).

ry commented on src/Link.tsx:
```diff
@@ -0,0 +1,15 @@
+));
+
+export const Link = (props) => {
+  if (props.to.indexOf("://")=== -1) {
```
Can you run "yarn format"?

josephsintum replied:
sure

ry commented:
can you also update the links elsewhere? For example there are many in src/Home

josephsintum commented:
> Looks good!
> 
> What do you think about also adding a `<Button>` component that uses MaterialButton? That way we get the same improvements for buttons too, not just links.
> 
> Oh and it would be cool if you could also add some tests :-).



> Looks good!
> 
> What do you think about also adding a `<Button>` component that uses MaterialButton? That way we get the same improvements for buttons too, not just links.
> 
> Oh and it would be cool if you could also add some tests :-).

I'll add <Button> to it

josephsintum commented:
I realized a lot people are using "href" or "to"
```tsx
<Link to="somewhere">some where</Link>
or 
<Link href="somewhere">some where</Link>
```

Should I enforce consistency by allow only <Link to=""> 
or allow flexibility with both

ry commented:
> Should I enforce consistency by allow only or allow flexibility with both

Consistency would be preferable. Maybe you can do that by defining an interface for the props.

josephsintum commented:
> > Should I enforce consistency by allow only or allow flexibility with both
> 
> Consistency would be preferable. Maybe you can do that by defining an interface for the props.

I have added the interface and added check to prevent "href" use

ry commented on src/components/Link.tsx:
```diff
@@ -0,0 +1,23 @@
+  /** Supports both internal and external links */
+  to: string;
+  /** Do not use href. Use <Link to="link" > instead for both internal and external links */
+  href?: never;
```
You can just leave this out.

josephsintum replied:
taking it out

ry commented on src/components/Link.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from "react";
```
Do not introduce a new directory - if you want to make a new organization it should be done in a separate PR.

ry commented on src/components/Link.tsx:
```diff
@@ -0,0 +1,23 @@
+  href?: never;
+}
+
+export const Link: React.FC<Props & LinkProps> = props => {
```
Maybe add some documentation here saying

```
/** Use this component instead of MaterialUI's Link or ReactRouter's Link. */
```

ry commented on src/Button.tsx:
```diff
@@ -0,0 +1,21 @@
+interface Props {
+  /** Supports both internal and external links */
+  to?: string;
+  component?: any;
```
s/any/React.Component/

josephsintum replied:
component?: HTMLElement;

lucacasonato replied:
@josephsintum HTMLElement is a default HTML element not a React component - you should use `React.ComponentType<any>`

lucacasonato commented on src/Button.tsx:
```diff
@@ -0,0 +1,22 @@
+}
+
+/** Use this component instead of MaterialUI's Button or ReactRouter's Link */
+export const Button: React.FC<Props & ButtonProps> = props => {
```
@ry What do you think of this? This just makes sure you don't use href by accident.
```suggestion
export const Button: React.FC<Props & Omit<ButtonProps, 'href'> = props => {
```
If this is changed make sure to change it [here](https://github.com/denoland/deno_website2/pull/91/files#diff-01e857eda18d514416246bb1ff1e7510R16) too.

ry replied:
Why can't we just use `<a href>`... I don't understand why getting a link is requiring so much infrastructure.

lucacasonato replied:
@ry https://github.com/denoland/deno_website2/issues/86#issuecomment-547166068

ry replied:
Thank you @lucacasonato - I missed that when you originally posted it. That was well put and very much helps me justify this.


@josephsintum - could you please copy @lucacasonato's comment into the top of src/Link.ts ? (since it is not at all obvious why this is necessary)

ry replied:
I don't understand why react-router's Link doesn't understand the difference between external and internal links tho.

josephsintum replied:
React doesn't handle routing 
So if `<a href>` is used, react will render the anchor tag without the SPA routing logic, then the whole page is reloaded which is not a great experience
so we use external packages like react-router which provides a `<Link to="...">` to handle routing, 
The link need to be styled by material ui, so react-router <Link> is wrapped with material ui <Link>

lucacasonato replied:
> I don't understand why react-router's Link doesn't understand the difference between external and internal links tho.

I agree - that is stupid, they should just handle external links as external links.

ry commented on src/Button.tsx:
```diff
@@ -0,0 +1,22 @@
+  let { to, ...rest } = props;
+  if (!props.to) {
+    return <MaterialButton {...props} />;
+  } else if (props.to.indexOf("://") === -1) {
```
this check should only be in one place in our code base.

ry commented:
LGTM - thanks @josephsintum and @lucacasonato
